### PR TITLE
libckteec: release active processing on functions failures

### PR DESCRIPTION
**Pull request by etienne-lms**
Change all impacted processing function to release the active processing
when it returns an error or an error other than CKR_BUFFER_TOO_SMALL.

Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>

### CODE REVIEW

**Comment by ruchi393** [comment]
Sorry for the late review. I somehow missed the PR.

`Reviewed-by: <Ruchika Gupta <ruchika.gupta@linaro.org>`

**Comment by etienne-lms** [comment]
tag applied

**Comment by vesajaaskelainen** [comment]
Sorry for being a bit late in response.

Original logic for _release active processing_ was that it will only be called when **input parameters** are wrong and require one to terminate active processing session.

Once we are sure that we can invoke the TA normally then we just invoke it and then that operation handles all necessary things to do.

Now what you have modified here is that when ever there is a failure it causes extra call to TA just to terminate while the one call could have done that which is not so nice. Making a TA call is somewhat expensive so lets try to keep that in minimum. So once `ckteec_invoke_ta()` is called one should not call `ck_release_active_processing()` -- instead the `ckteec_invoke_ta()` should terminate the processing during its own internal operations.

Other thing that was lost is logical descriptions for argument handling which if you think should be removed then fine.

File: libckteec/src/pkcs11_processing.c [hunk 1]
```diff
@@ -174,8 +174,10 @@ CK_RV ck_encdecrypt_update(CK_SESSION_HANDLE session,
 	uint32_t session_handle = session;
 	size_t out_size = 0;
 
-	if ((out_len && *out_len && !out) || (in_len && !in))
-		return CKR_ARGUMENTS_BAD;
+	if ((out_len && *out_len && !out) || (in_len && !in)) {
+		rv = CKR_ARGUMENTS_BAD;
+		goto bail;
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:179]: `A call to C_EncryptUpdate which results in an error other than CKR_BUFFER_TOO_SMALL terminates the current encryption operation.`

So from logical point of view it is correct to terminate active processing in here with invalid arguments.

Also it is correct to terminate for `CKR_HOST_MEMORY`.

File: libckteec/src/pkcs11_processing.c [hunk 2]
```diff
@@ -223,6 +225,11 @@ CK_RV ck_encdecrypt_update(CK_SESSION_HANDLE session,
 	ckteec_free_shm(in_shm);
 	ckteec_free_shm(ctrl);
 
+	if (rv != CKR_OK && rv != CKR_BUFFER_TOO_SMALL)
+		ck_release_active_processing(session, decrypt ?
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:229]: This should not be invoked after `ckteec_invoke_ta()`.

File: libckteec/src/pkcs11_processing.c [hunk 3]
```diff
@@ -240,8 +247,10 @@ CK_RV ck_encdecrypt_oneshot(CK_SESSION_HANDLE session,
 	uint32_t session_handle = session;
 	size_t out_size = 0;
 
-	if ((out_len && *out_len && !out) || (in_len && !in))
-		return CKR_ARGUMENTS_BAD;
+	if ((out_len && *out_len && !out) || (in_len && !in)) {
+		rv = CKR_ARGUMENTS_BAD;
+		goto bail;
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:252]: `A call to C_Encrypt always terminates the active encryption operation unless it returns CKR_BUFFER_TOO_SMALL or is a successful call (i.e., one which returns CKR_OK) to determine the length of the buffer needed to hold the ciphertext.`

So from logical point of view it is correct to terminate active processing in here with invalid arguments.

Also it is correct to terminate for CKR_HOST_MEMORY.

File: libckteec/src/pkcs11_processing.c [hunk 4]
```diff
@@ -286,6 +295,11 @@ CK_RV ck_encdecrypt_oneshot(CK_SESSION_HANDLE session,
 	ckteec_free_shm(in_shm);
 	ckteec_free_shm(ctrl);
 
+	if (rv != CKR_OK && rv != CKR_BUFFER_TOO_SMALL)
+		ck_release_active_processing(session, decrypt ?
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:299]: This should not be invoked after `ckteec_invoke_ta()`.

File: libckteec/src/pkcs11_processing.c [hunk 5]
```diff
@@ -300,8 +314,10 @@ CK_RV ck_encdecrypt_final(CK_SESSION_HANDLE session,
 	uint32_t session_handle = session;
 	size_t out_size = 0;
 
-	if (out_len && *out_len && !out)
-		return CKR_ARGUMENTS_BAD;
+	if (out_len && *out_len && !out) {
+		rv = CKR_ARGUMENTS_BAD;
+		goto bail;
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:319]: `A call to C_EncryptFinal always terminates the active encryption operation unless it returns CKR_BUFFER_TOO_SMALL or is a successful call (i.e., one which returns CKR_OK) to determine the length of the buffer needed to hold the ciphertext.`

So from logical point of view it is correct to terminate active processing in here with invalid arguments.

Also it is correct to terminate for CKR_HOST_MEMORY.

File: libckteec/src/pkcs11_processing.c [hunk 6]
```diff
@@ -338,6 +354,11 @@ CK_RV ck_encdecrypt_final(CK_SESSION_HANDLE session,
 	ckteec_free_shm(out_shm);
 	ckteec_free_shm(ctrl);
 
+	if (rv != CKR_OK && rv != CKR_BUFFER_TOO_SMALL)
+		ck_release_active_processing(session, decrypt ?
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:358]: This should not be invoked after `ckteec_invoke_ta()`.

File: libckteec/src/pkcs11_processing.c [hunk 7]
```diff
@@ -461,6 +476,9 @@ CK_RV ck_digest_update(CK_SESSION_HANDLE session, CK_BYTE_PTR in, CK_ULONG in_le
 	ckteec_free_shm(io1);
 	ckteec_free_shm(ctrl);
 
+	if (rv)
+		ck_release_active_processing(session, PKCS11_CMD_DIGEST_UPDATE);
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:480]: This should not be invoked after `ckteec_invoke_ta()`.

File: libckteec/src/pkcs11_processing.c [hunk 8]
```diff
@@ -551,6 +548,10 @@ CK_RV ck_digest_oneshot(CK_SESSION_HANDLE session, CK_BYTE_PTR in,
 	ckteec_free_shm(io2);
 	ckteec_free_shm(ctrl);
 
+	if (rv != CKR_OK && rv != CKR_BUFFER_TOO_SMALL)
+		ck_release_active_processing(session,
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:552]: This should not be invoked after `ckteec_invoke_ta()`.

File: libckteec/src/pkcs11_processing.c [hunk 9]
```diff
@@ -613,6 +614,9 @@ CK_RV ck_digest_final(CK_SESSION_HANDLE session, CK_BYTE_PTR digest_ref,
 	ckteec_free_shm(io2);
 	ckteec_free_shm(ctrl);
 
+	if (rv != CKR_OK && rv != CKR_BUFFER_TOO_SMALL)
+		ck_release_active_processing(session, PKCS11_CMD_DIGEST_FINAL);
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:618]: This should not be invoked after `ckteec_invoke_ta()`.

File: libckteec/src/pkcs11_processing.c [hunk 10]
```diff
@@ -703,6 +709,10 @@ CK_RV ck_signverify_update(CK_SESSION_HANDLE session,
 	ckteec_free_shm(in_shm);
 	ckteec_free_shm(ctrl);
 
+	if (rv)
+		ck_release_active_processing(session, sign ?
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:713]: This should not be invoked after `ckteec_invoke_ctrl_in()`.

File: libckteec/src/pkcs11_processing.c [hunk 11]
```diff
@@ -774,6 +786,11 @@ CK_RV ck_signverify_oneshot(CK_SESSION_HANDLE session,
 	ckteec_free_shm(io2);
 	ckteec_free_shm(ctrl);
 
+	if (rv != CKR_OK && rv != CKR_BUFFER_TOO_SMALL)
+		ck_release_active_processing(session, sign ?
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:790]: This should not be invoked after `ckteec_invoke_ta()`.

File: libckteec/src/pkcs11_processing.c [hunk 12]
```diff
@@ -829,6 +848,11 @@ CK_RV ck_signverify_final(CK_SESSION_HANDLE session,
 	ckteec_free_shm(io);
 	ckteec_free_shm(ctrl);
 
+	if (rv != CKR_OK && rv != CKR_BUFFER_TOO_SMALL)
+		ck_release_active_processing(session, sign ?
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:852]: This should not be invoked after `ckteec_invoke_ta()`.

File: libckteec/src/pkcs11_processing.c [hunk 13]
```diff
@@ -721,8 +731,10 @@ CK_RV ck_signverify_oneshot(CK_SESSION_HANDLE session,
 	size_t out_size = 0;
 
 	if ((in_len && !in) || (sign_len && *sign_len && !sign_ref) ||
-	    (sign && !sign_len))
-		return CKR_ARGUMENTS_BAD;
+	    (sign && !sign_len)) {
+		rv = CKR_ARGUMENTS_BAD;
+		goto bail;
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:736]: `A call to C_Sign always terminates the active signing operation unless it returns CKR_BUFFER_TOO_SMALL or is a successful call (i.e., one which returns CKR_OK) to determine the length of the buffer needed to hold the signature.`

So from logical point of view it is correct to terminate active processing in here with invalid arguments.

Also it is correct to terminate for CKR_HOST_MEMORY.

File: libckteec/src/pkcs11_processing.c [hunk 14]
```diff
@@ -788,8 +805,10 @@ CK_RV ck_signverify_final(CK_SESSION_HANDLE session,
 	uint32_t session_handle = session;
 	size_t io_size = 0;
 
-	if ((sign_len && *sign_len && !sign_ref) || (sign && !sign_len))
-		return CKR_ARGUMENTS_BAD;
+	if ((sign_len && *sign_len && !sign_ref) || (sign && !sign_len)) {
+		rv = CKR_ARGUMENTS_BAD;
+		goto bail;
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:810]: `A call to C_SignFinal always terminates the active signing operation unless it returns CKR_BUFFER_TOO_SMALL or is a successful call (i.e., one which returns CKR_OK) to determine the length of the buffer needed to hold the signature.`

So from logical point of view it is correct to terminate active processing in here with invalid arguments.

Also it is correct to terminate for CKR_HOST_MEMORY.

File: libckteec/src/pkcs11_processing.c [hunk 15]
```diff
@@ -678,8 +682,10 @@ CK_RV ck_signverify_update(CK_SESSION_HANDLE session,
 	TEEC_SharedMemory *in_shm = NULL;
 	uint32_t session_handle = session;
 
-	if (in_len && !in)
-		return CKR_ARGUMENTS_BAD;
+	if (!in && in_len) {
+		rv = CKR_ARGUMENTS_BAD;
+		goto bail;
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:687]: `A call to C_SignUpdate which results in an error terminates the current signature operation.`

So from logical point of view it is correct to terminate active processing in here with invalid arguments.

Also it is correct to terminate for CKR_HOST_MEMORY.

File: libckteec/src/pkcs11_processing.c [hunk 16]
```diff
@@ -572,8 +573,8 @@ CK_RV ck_digest_final(CK_SESSION_HANDLE session, CK_BYTE_PTR digest_ref,
 	 * to get size of the required buffer
 	 */
 	if (!digest_len) {
-		ck_release_active_processing(session, PKCS11_CMD_DIGEST_FINAL);
-		return CKR_ARGUMENTS_BAD;
+		rv = CKR_ARGUMENTS_BAD;
+		goto bail;
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:577]: From logical point of view it is correct to terminate active processing in here with invalid arguments.

Also it is correct to terminate for CKR_HOST_MEMORY which was missed originally.

File: libckteec/src/pkcs11_processing.c [hunk 17]
```diff
@@ -476,30 +494,9 @@ CK_RV ck_digest_oneshot(CK_SESSION_HANDLE session, CK_BYTE_PTR in,
...
-	 * - If digest_ref == NULL BUT digest_len != NULL just operate normally
-	 * to get size of the required buffer
-	 */
-	if (!digest_len) {
-		ck_release_active_processing(session,
-					     PKCS11_CMD_DIGEST_ONESHOT);
-		return CKR_ARGUMENTS_BAD;
+	if ((!in && in_len) || !digest_len) {
+		rv = CKR_ARGUMENTS_BAD;
+		goto bail;
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:499]: From logical point of view it is correct to terminate active processing in here with invalid arguments.

Also it is correct to terminate for CKR_HOST_MEMORY which was missed originally.

File: libckteec/src/pkcs11_processing.c [hunk 18]
```diff
@@ -418,24 +439,18 @@ CK_RV ck_digest_key(CK_SESSION_HANDLE session, CK_OBJECT_HANDLE key)
...
-	 * - If in == NULL AND in_len != 0 -> need to call TA to
-	 * terminate session.
-	 * - If in == NULL AND in_len == 0 -> just operate normally.
-	 */
 	if (!in && in_len) {
-		ck_release_active_processing(session,
-					     PKCS11_CMD_DIGEST_UPDATE);
-		return CKR_ARGUMENTS_BAD;
+		rv = CKR_ARGUMENTS_BAD;
+		goto bail;
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:453]: From logical point of view it is correct to terminate active processing in here with invalid arguments.

Also it is correct to terminate for CKR_HOST_MEMORY which was missed originally.

File: libckteec/src/pkcs11_processing.c (hunk 2 above)
Comment (etienne-lms) [reply libckteec/src/pkcs11_processing.c:229]: agree

File: libckteec/src/pkcs11_processing.c (hunk 4 above)
Comment (etienne-lms) [reply libckteec/src/pkcs11_processing.c:299]: agree

**Comment by etienne-lms** [comment]
I agree with the comments. Please see the fixup commit.

**Comment by etienne-lms** [comment]
> Other thing that was lost is logical descriptions for argument handling which if you think should be removed then fine but then remove them from all places. For today's compilers generated code for both versions should be equal so now only "optimization" is lost comments.

I didn't want to copy/paste inline comment's stating why we release the processing. The arguments are bad and that's all. For consistency I removed the extra description.

If you feel the descriptions should be back, I'm fine. @ruchi393, your thoughts?

**Comment by etienne-lms** [comment]
Thanks @vesajaaskelainen for the feedback and details. I hope it's all fixed.

File: libckteec/src/pkcs11_processing.c [hunk 19]
```diff
@@ -218,7 +220,14 @@ CK_RV ck_encdecrypt_update(CK_SESSION_HANDLE session,
 	if (rv == CKR_BUFFER_TOO_SMALL && out_size && !out)
 		rv = CKR_OK;
 
+	goto bail_invoked;
+
 bail:
+	if (rv != CKR_OK && rv != CKR_BUFFER_TOO_SMALL)
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:226]: I suppose this if is always true -- applies to all similar things.

**Comment by vesajaaskelainen** [comment]
> Thanks @vesajaaskelainen for the feedback and details. I hope it's all fixed.

There is one comment with `if` clauses -- with or without it:
`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

File: libckteec/src/pkcs11_processing.c (hunk 19 above)
Comment (etienne-lms) [reply libckteec/src/pkcs11_processing.c:226]: :laughing:  good catch!

**Comment by etienne-lms** [comment]
@vesajaaskelainen, i've addressed you comment in latest fixup commit.
@vesajaaskelainen & @ruchi393, could you please confirm your review tag?

**Comment by ruchi393** [comment]
LGTM. My R-by tag still applies.

**Comment by vesajaaskelainen** [comment]
> @vesajaaskelainen, i've addressed you comment in latest fixup commit.
> @vesajaaskelainen & @ruchi393, could you please confirm your review tag?

Looks good.

**Comment by etienne-lms** [comment]
Tags applied.

**Pull request merged**